src/main/java/org/opensearch/security/user/User.java

```diff
@@ -272,13 +272,13 @@ public synchronized final Map<String, String> getCustomAttributesMap() {
         return attributes;
     }
 
-    public final void addSecurityRoles(final Collection<String> securityRoles) {
+    public synchronized final void addSecurityRoles(final Collection<String> securityRoles) {
```

Comment:
ConcurrentModificationException is thrown when iterating through a collection its contents are changed - so the iterator is invalid [1].  In `PrivilegesEvaluator` we are still iterating over the collection - outside of a synchronized block were the iterator could be modified.

To prevent this exception we would need to remove the `addSecurityRoles(...)` function, because this it updates the underlying collection - @cwperks calls this out as well.  

Alternatively - less clean, change addSecurityRoles(...) so it changes the collection after its been updated, ala change the field `securityRoles` to be an ActomicReference<Set<String>>, then do `securityRole.set(Sets.union(securityRoles.get(), securityRoles)`, if any references to the original securityRoles exist in another thread/memory they are different references than the new object that was created


- [1] https://stackoverflow.com/questions/602636/why-is-a-concurrentmodificationexception-thrown-and-how-to-debug-it

Reply:
@peternied and @cwperks, Yes, I agree. Working on removing this `addSecurityRoles()`